lib/Onnxifi/Base.cpp

```diff
@@ -43,6 +47,53 @@ void Event::wait() {
   cond_.wait(guard, [this] { return fired_ == true; });
 }
 
+constexpr std::chrono::microseconds ThreadPool::kThreadBreakDuration;
+
+ThreadPool::ThreadPool(unsigned numWorkers) : shouldStop_(false) {
+  // Intialize all workers and make each one run threadPoolWorkerMain.
+  for (unsigned i = 0; i < numWorkers; i++) {
+    std::thread th(std::bind(&ThreadPool::threadPoolWorkerMain, this));
+    workers_.push_back(std::move(th));
+  }
+}
+
+ThreadPool::~ThreadPool() {
+  // Signal to workers to stop.
+  shouldStop_ = true;
+
+  // Join all worker threads.
+  for (auto &w : workers_) {
+    w.join();
+  }
+  workers_.clear();
+}
+
+void ThreadPool::submit(const std::function<void(void)> &fn) {
+  // Add fn to the work queue.
+  std::lock_guard<std::mutex> lock(workQueueMtx_);
+  workQueue_.push(std::move(fn));
+}
+
+void ThreadPool::threadPoolWorkerMain() {
+  std::unique_lock<std::mutex> lock(workQueueMtx_, std::defer_lock);
+
+  while (!shouldStop_) {
+    lock.lock();
+
+    if (!workQueue_.empty()) {
+      // Pop and process a work item.
+      auto workItem = std::move(workQueue_.front());
+      workQueue_.pop();
+      lock.unlock();
+      workItem();
+    } else {
+      // Wait for a while and then check the queue again.
+      lock.unlock();
+      std::this_thread::sleep_for(kThreadBreakDuration);
```

Comment:
need proper signaling mechanism when an item is added to the queue.

once you extract this class into a separate file, please, add a test

Reply:
I meant to use cond_var for example to signal when the queue is not empty.

Reply:
Oh, I see.

Reply:
Here is an example thread pool: 
https://github.com/pytorch/pytorch/blob/dd4b9b06a434ca3b72c8a091c8739c33325ccead/caffe2/utils/thread_pool.h#L24-L180

Note that the overhead of synchronization is pretty high if we just use `mutex` and `conditional_variable`. High performance usually requires a better concurrent design but I think this might be a good first implementation.